### PR TITLE
Deduplicate similar regexps in the router to reduce memory retention

### PR DESCRIPTION
**Pull request by casperisfine**
While profiling our app boot, I noticed that journey was rebuilding the same small set of regexps a lot of times:

```
Allocated String Report
-----------------------------------
 663.88 kB   16597  "\\A(?-mix:[^.\\/?]+)\\Z"
             16592  rails-c8f0c6450c75/actionpack/lib/action_dispatch/journey/gtg/transition_table.rb:168
                 5  rails-c8f0c6450c75/actionpack/lib/action_dispatch/journey/gtg/transition_table.rb:14

...

 331.92 kB    8298  "(?-mix:[^.\\/?]+)"
              8296  rails-c8f0c6450c75/actionpack/lib/action_dispatch/journey/gtg/transition_table.rb:168
                 2  rails-c8f0c6450c75/actionpack/lib/action_dispatch/journey/gtg/transition_table.rb:14

...

 445.63 kB    5064  "(?-mix:\\b(\\d{4}-\\d{2}|unstable|unversioned)\\b)"
              3372  rails-c8f0c6450c75/actionpack/lib/action_dispatch/journey/path/pattern.rb:206
              1686  rails-c8f0c6450c75/actionpack/lib/action_dispatch/journey/path/pattern.rb:99
                 6  rails-c8f0c6450c75/actionpack/lib/action_dispatch/journey/gtg/transition_table.rb:168
```

The profiler can only display strings, but they are then used to build the corresponding regexps. Even worse in one case the pattern is `Regexp.new(/\A#{source}\Z/)`, which means we build two regexps for no reason.

Regexps are costly in tow ways: they use quite a lot of memory, and they also take quite a bit of CPU to compile when they are instantiated.

So I believe we can save a bit of resident memory by deduplicating these regexps. Of course the saved amount depends on the number of routes. 

In our case the saving is about `2.85MiB`.

Before:
```
retained memory by class
-----------------------------------
  10.87 MB  Regexp
```

After:
```
retained memory by class
-----------------------------------
   7.50 MB  Regexp
```

### Implementations

I left two commits because I have two possible implementations and I'm unsure which I should favor.

The first one use a regular hash, so we can use equality to deduplicate the regexp. The problem is that it might cause us to hold onto objects that would normally be GCed (`~140KiB` more in our app).

The second one uses `ObjectSpace::WeakMap`, so the strings have to be interned before they can be used as keys, because `WeakMap` use identity comparison.

Ideally we could use the simpler implementation, and clear the hash when we know we won't need it anymore, but I couldn't find a proper hook for that. If someone has an idea about that.

cc @rafaelfranca thoughts?

### CODE REVIEW

**Comment by jhawthorn** [comment]
This is a good idea, we see a ton of memory usage here as well.

What about making this optimization to Ruby instead. I see that's a [thought you've already had 😁](https://bugs.ruby-lang.org/issues/16557).

> The second one uses `ObjectSpace::WeakMap`, so the strings have to be interned before they can be used as keys, because `WeakMap` use identity comparison.
> 
> Ideally we could use the simpler implementation, and clear the hash when we know we won't need it anymore, but I couldn't find a proper hook for that. If someone has an idea about that.

Maybe we could be storing the hash/cache on the route set or a similar object?

**Comment by casperisfine** [comment]
> What about making this optimization to Ruby instead. I see that's a [thought you've already had 😁](https://bugs.ruby-lang.org/issues/16557).

Ahah good catch. Well, this Ruby side solution would only concern Regexp literals, here they are built dynamically, so it's kind of orthogonal anyway. Also I never managed to get that patch working, as I ran into some GC issue and never managed to get someone to help me figure it out. So I wouldn't count on it too much, but I'd really love to finish it one day. 

> Maybe we could be storing the hash/cache on the route set or a similar object?

My problem is not so much where it is stored, but to find a proper place to clear it. I tried `config.after_initialize { if config.eagerload }`, but it wasn't late enough. I'm sure there's a proper place though.

**Comment by jhawthorn** [comment]
cc @composerinteralia

**Comment by casperisfine** [comment]
I think I'll just merge the WeakMap version for now.

**Comment by composerinteralia** [comment]
Of the three specific strings you mentioned in your PR description, it looks like two are the DEFAULT_EXP `/[^.\/?]+/`. Could we solve this by adding some special case handling of that particular regexp?

For the third one, I'm not seeing that this PR addresses it. It appears to be coming from the Regexp.union call in https://github.com/rails/rails/blob/c8f0c6450c758dcc8e6894f7c636ec4d2692b773/actionpack/lib/action_dispatch/journey/path/pattern.rb#L206. Could we fix that by only calling union when necessary? Something like:

```diff
-                re = /#{Regexp.union(@requirements[node])}|/
+                requirement = @requirements[node]
+                requirement = Regexp.union(requirement) if Array === requirement
+                re = /#{requirement}|/
```

(The purpose of this code is to count how many capture groups the regex has. Maybe that's obvious to others, but it took me a minute to figure out.)

**Comment by casperisfine** [comment]
> Could we solve this by adding some special case handling of that particular regexp?

Hum, interesting proposition. I suppose it would solve the issue better for the vast majority of users, but I think I'd need to collect some more data to know wether that solves the larger problem for us (we have loads of routes).

> For the third one, I'm not seeing that this PR addresses it

Indeed, I haven't looked at it yet.  Your suggestion make sense to me.

**Comment by casperisfine** [comment]
> Could we solve this by adding some special case handling of that particular regexp?

So I computed the list of unique regexps we get in our routes

```ruby
4162 /[^.\/?]+/
  42 /(js|json|html)/
  34 /json/
  30 /.+?/
   8 /\d{4}/
   6 /(json|xml)/
   4 /[^\/]+/
   4 /.*/
   4 /([a-z0-9]+(:|%3A)\d+(,|%2C)?)+/
   4 /(?!api)([a-zA-Z]{2,3})(?:-([a-zA-Z0-9]+))?/
   3 /c|os|co|o|ac|cn/
   3 /\d+/
   3 /\b(\d{4}-\d{2}|unstable|unversioned)\b/
   3 /(.+)/
   2 /ping-api(-staging)*/
   2 /messaging(-staging)*/
   2 /\b(\d{4}-\d{2}|unstable)\b/
   2 /[a-f0-9]{32}/
   2 /[\w+\.]+/
   2 /.+/
   2 /(a|apps|community|tools)/
   1 /ca3-.*/
   1 /[a-zA-Z0-9\.-]+/
   1 /(html|json|xml)/```

So I think you are right, by just handling the default one, we'd solve most of the problem. Of course some users that rarely use the default wouldn't benefit from it, but I think it's a good tradeoff. I'll open a PR with just that change, and will take another look at the other callsites after.

File: actionpack/lib/action_dispatch/journey/path/pattern.rb [hunk 1]
```diff
@@ -83,7 +83,7 @@ def initialize(separator, matchers)
           end
 
           def accept(node)
-            %r{\A#{visit node}\Z}
+            RegexpCache.anchored(visit(node))
```
Comment (casperisfine) [inline actionpack/lib/action_dispatch/journey/path/pattern.rb:86]: Here it's much less clear. I count `4771` regexp, `3573` different ones, and the most duplicated one is seen 8 times, and only `698` of them are only seen once.

So not sure if it's really worth it.

**Comment by casperisfine** [comment]
> Could we fix that by only calling union when necessary?

I think so, but that third place doesn't hold on the regexp, hence why I didn't really focus on it. From our stackprof reports that represent 0.2% of our boot time, if there's an easy fix it might be worth it, but nothing really crazy.

**Comment by composerinteralia** [comment]
> that represent 0.2% of our boot time

Yup, agree it's probably not worth doing much there.

File: actionpack/lib/action_dispatch/journey/path/pattern.rb (hunk 1 above)
Comment (composerinteralia) [reply actionpack/lib/action_dispatch/journey/path/pattern.rb:86]: Yeah, it makes sense these would be mostly all different. I think `visit node` here returns a regex source (i.e. a string) for the whole path. So there we'd get one regexp for each route.

We've got `4407` regexp (and `4407` routes), `3720` different ones, and `3220` of them only appear once. One of them appears 10 times.

So not really worth it for us, although it seems like you've got more routes with the same regex than we do.

**Pull request closed without merging**